### PR TITLE
[PLAT-5291] Refactor pipeline to use multistage builds

### PR DESCRIPTION
**Pull request by Cawllec**
## Goal

Initial refactor to using dynamic pipelines to allow quick "smoke" tests to run vs "full" tests.

Smoke tests are triggered by default, or if the commit message contains `[quick ci]`
Full tests are triggered if the commit message contains `[full ci]`, or if the branch running is, or is merging into, `master` or `next`

Aiming this into an integration branch as I need to add more to the smoke tests before we have this present generally.

## Design

Most of the common steps are still present in `pipeline.yml`, with 3 main exceptions:
- Maze runner container build step
- Integration test steps
- E2E steps

These are triggered slightly differently depending on whether it's a smoke or full test set:
- The maze runner container has the `full_tests` or `smoke_tests` feature set added depending on the pipeline.
- Only NDK r16b integration test steps are run on the smoke tests
- The E2E maze-runner steps are run based on the maze-runner container previously built.  In the future we may decide which Android versions or Android/NDK combinations to target with smoke vs full test sets.

## Testing

[An example of a smoke test run](https://buildkite.com/bugsnag/bugsnag-android/builds/2668)
[An example of a full test run](https://buildkite.com/bugsnag/bugsnag-android/builds/2670#_)

### CODE REVIEW

File: dockerfiles/Dockerfile.android-maze-runner
Comment (fractalwrench) [inline dockerfiles/Dockerfile.android-maze-runner:8]: Is it worth just always copying all the tests, or would there be some side effect to that which is undesirable?

It would definitely be good to run the `_smoke_test.feature` on the full CI run either way - as that should get quick feedback about whether a feature is successful when running against `next/master`.

File: .buildkite/full_pipeline_trigger.sh
Comment (fractalwrench) [inline .buildkite/full_pipeline_trigger.sh:3]: Do we want this to be possible? It would allow developers to skip the full checks on next/master.

File: .buildkite/pipeline.smoke.yml
Comment (fractalwrench) [inline .buildkite/pipeline.smoke.yml:11]: Running the instrumentation tests on just 16b + 4.4 would be fine for a quick CI run - there's no advantage from my perspective in running these tests 4x.

File: .buildkite/pipeline.full.yml
Comment (fractalwrench) [inline .buildkite/pipeline.full.yml:61]: An existing issue, but this appears to be a dupe job

File: .buildkite/pipeline.full.yml
Comment (fractalwrench) [inline .buildkite/pipeline.full.yml:9]: Would it be possible to tweak the ordering of jobs so that one mazerunner job runs after the first instrumentation test? That might help speed up feedback when Browserstack is saturated

**Review by fractalwrench — COMMENTED** [review]
Looks like this will be fast :)

This looks like the right approach. I had a few queries which are mostly around what the process for quick/full CI runs should be - I've left these as inline comments.

**Comment by twometresteve** [comment]
Please always name integration branches with a `integration/` prefix.  It's then clear to all of its purpose and it also allows us to use a single branch protection rule to cover all integration branches (by setting it on "integration/*".

File: dockerfiles/Dockerfile.android-maze-runner
Comment (Cawllec) [reply dockerfiles/Dockerfile.android-maze-runner:8]: We could have a single maze-runner container and leave the individual config to the step in the pipeline, I'm not opposed to that

File: .buildkite/full_pipeline_trigger.sh
Comment (Cawllec) [reply .buildkite/full_pipeline_trigger.sh:3]: Possibly not, just added it as a utility.

File: .buildkite/full_pipeline_trigger.sh
Comment (twometresteve) [inline .buildkite/full_pipeline_trigger.sh:7]: We'll also want full ci if it's a PR into `next` or `master`.  I think you can use `BUILDKITE_PULL_REQUEST_BASE_BRANCH` for that.

File: .buildkite/pipeline.full.yml
Comment (twometresteve) [inline .buildkite/pipeline.full.yml:224]: These should come out for Android 6 and 7 now.

File: dockerfiles/Dockerfile.android-maze-runner
Comment (Cawllec) [reply dockerfiles/Dockerfile.android-maze-runner:7]: Ok, we can adjust to that when necessary

File: dockerfiles/Dockerfile.android-maze-runner
Comment (twometresteve) [reply dockerfiles/Dockerfile.android-maze-runner:8]: In the Cocoa pipeline we consume the Maze Runner image directly in `docker-compose.yml` and mount `features/` as a volume.  There is no build step in the pipeline at all - would that be an option here?  Although it takes next to no time to build the image, there is instance wait time and general use to consider.  We could have an annoying wait time to get a fresh agent, just for it to then sit idle for 30 minutes before being surrendered.

File: dockerfiles/Dockerfile.android-maze-runner
Comment (twometresteve) [reply dockerfiles/Dockerfile.android-maze-runner:7]: Not a strong preference by any means, but I like the one-word descriptions and matching the feature directory name to the commit message term would seem to make sense.  If I were to choose I'd probably say full, quick and minimal here.

File: .buildkite/full_pipeline_trigger.sh
Comment (twometresteve) [reply .buildkite/full_pipeline_trigger.sh:3]: Hmm, slightly in two minds here as Github status checks should cover us.  I'm not sure how much we'd actually want to use it and if there is a compelling case for it.

File: .buildkite/full_pipeline_trigger.sh
Comment (twometresteve) [inline .buildkite/full_pipeline_trigger.sh:5]: Could we make the commit string lower case, i.e. `[full ci]`?  It would then be consistent with Buidlkite's `[skip ci]` construct and I think a little easier to type.

Comment (twometresteve) [reply .buildkite/full_pipeline_trigger.sh:5]: ...or do a case-insensitive string comparison, of course.

File: dockerfiles/Dockerfile.android-maze-runner
Comment (Cawllec) [reply dockerfiles/Dockerfile.android-maze-runner:8]: Yes, that would make a lot more sense

File: .buildkite/full_pipeline_trigger.sh
Comment (Cawllec) [reply .buildkite/full_pipeline_trigger.sh]: Just thought we have cases where a changelog gets changed or typo gets fixed before a release and we often have to wait for a full CI run for that minor change.  Would it be acceptable to use quick-ci in these cases?

**Review by fractalwrench — APPROVED** [review]

File: .buildkite/full_pipeline_trigger.sh
Comment (twometresteve) [reply .buildkite/full_pipeline_trigger.sh]: With a new commit the Github status checks will be reset, so we'd need another full run _at some point_ before being able to merge - but perhaps I can see the use...  If there are a few review comments to action (including code changes) and the author wants to address them in separate commits with fast-feedback, it could be useful to be able to push them with `[quick ci]` in the knowledge that a full run will be needed before being able to actually merge (which can then be achieved by simply amending the last commit message).  Let's leave it in for now and see how it pans out.

**Review by twometresteve — APPROVED** [review]
LGTM for this stage of the work.

**Pull request merged**
